rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,238 @@
+understandable format). Additional benefits of configuration-based node creation
+include:
+
+- It would be easier to build tools to visualise the node graph without running
+  the application.
+- It would be easier to build tools to perform static analysis of the node graph
+  to detect potential problems.
+- It would usually be easier to reason about the security of a static node
+  graph.
+
+## Overview
+
+Moving all node and channel creation to a configuration-based mechanism is a
+significant and potentially disruptive change. This RFC proposes a more gradual
+approach by implementing a re-usable Wasm-based planner node that uses a
+manifest file to do the node and channel creation. The functionality can be
+expanded over time to support more complex scenarios.
+
+The planner node will be responsible for creating nodes, creating channels,
+sending channel handles to the created nodes to ensure that the full node graph
+is created on start-up. It will will also act as a message router, which would
+route HTTP- and GRPC invocations to other nodes based on the associated labels.
+
+## Detailed Design
+
+### Planner Node
+
+The planner node will the the initial node created for an application. The
+manifest file will be passed to the planner node as part of the application's
+`ConfigMap`.
+
+The planner node will create all the nodes and channels as specified with the
+appropriate labels. It will send initialisation messages to each of the nodes,
+providing additional configuration and handles to channels. It is the
+responsibility of each of the nodes to store these channels for future
+communications with other nodes. It will send a filtered version of the config
+map as part of the initialisation message, based on the configuration keys
+specified in the `config_files` field.
+
+### Invocation Routing
+
+The planner node will also be responsible for routing invocations from the gRPC-
+and HTTP server pseudo nodes to the right receiver node based on the associated
+labels. Since it would be possible for multiple nodes to have the same label,
+the planner node would need an indication of which node should be the invocation
+recipient for a specific label.
+
+### Manifest File
+
+The manifest file will be will be passed to the application as part of the
+initial configuration map as a TOML file. The file will be deserialised into the
+following structure:
+
+```rust
+
+struct Node {
+  /// Unique name for the node.
+  name: String,
+
+  /// The label associated with the node.
+  label: Label,
+
+  /// The keys of the files from the config map that must be passed to the node
+  /// as part of the initialisation message. This will be used to generate a
+  /// filtered view of the map that will be sent as part of the initialisation
+  /// message.
+  config_files: Vec<String>,
+
+  /// Whether this node will accept gRPC invocations targeted at the specified
+  /// label.
+  grpc_invocation_receiver: bool,
```

Comment:
This seems very specific, is the idea that there is special logic that looks at a label and dispatches to this node based on the label? What if more than one node has this bool set to true? What if data needs to be pre-processed before being sent to the node (e.g. in the PIR example that @ipetr0v is working on, in which a handle to a channel containing the public database is passed alongside each invocation )? I think this is too limiting, conceptually it seems similar to a type system that only allows dynamic dispatching based on a specific input message, but it seems a special case rather than a generic mechanism.

Reply:
The idea is that the planner node is also the default router. This is very specific to deal with the current situation where each gRPC and HTTP pseudo node only has one channel over which it sends all invocations. Part of the validation of the manifest should be that only one node per label has this value set to true.

In the case where additional processing needs to be done (e.g. the trusted database example) there are two options. This first is to pass the database contents as part of the confg map to the dynamically created node in the intitialisation message. The other is to support specifying a different public node to do the routing. The planner node can then forward the invocation receiver handle to the custom node.

Reply:
This idea is based on passing the invocations to the nodes over the initialisation channel that the planner node already has open to each node.

A different option to address this in a more generic way is to specify the proto message for each channel. We could then set up a channel for either `GrpcInvocation` or `HttpInvocation` with the planner node as the sender node and the invocation target as the receiver node. The planner can then validate that there is only one such channel for each label. 

If the planner node receives an invocation with a label (on the request receiver) for which it does not have a channel, it can return an error on the response sender so the client is informed that the request could not be processed.

Similarly, if more advanced routing is required, we can create a channel for GrpcInvocationReceiver (or HttpInvocationReceiver) messages. If the planner node sees such a channel, it can forward the receiver handle for the invocations to that node and no longer act as the router.